src/equations/serre_green_naghdi_1d.jl

```diff
@@ -911,6 +1020,156 @@ function rhs_sgn_upwind!(dq, q, t, equations, source_terms, solver, cache,
     return nothing
 end
 
+# flat bathymetry with reflecting boundary conditions
+function create_cache(mesh,
+                      equations::SerreGreenNaghdiEquations1D{BathymetryFlat},
+                      solver,
+                      initial_condition,
+                      ::BoundaryConditionReflecting,
+                      RealT, uEltype)
+    (; D1, D2) = solver
+
+    # create temporary storage
+    h = ones(RealT, nnodes(mesh))
+    b = zero(h)
+    h_x = zero(h)
+    v_x = zero(h)
+    h2_x = zero(h)
+    hv_x = zero(h)
+    v2_x = zero(h)
+    h2_v_vx_x = zero(h)
+    h_vx_x = zero(h)
+    p_x = zero(h)
+    tmp = zero(h)
+    M_h = zero(h)
+
+    if D1 isa DerivativeOperator && D2 isa VarCoefDerivativeOperator
+        A = BandedMatrix(D2)
+        cache = (; h, b, h_x, v_x, h2_x, hv_x, v2_x,
+                 h2_v_vx_x, h_vx_x, p_x, tmp,
+                 M_h, A,
+                 D1, D2)
+    else
+        throw(ArgumentError("Combination of operators not supported."))
+    end
+
+    return cache
+end
+
+function assemble_system_matrix!(cache, h, D1, D2::VarCoefDerivativeOperator,
+                                 ::SerreGreenNaghdiEquations1D{BathymetryFlat})
+    (; M_h, A) = cache
+
+    @.. M_h = h
+    scale_by_mass_matrix!(M_h, D1)
+
+    # Floating point errors accumulate a bit and the system matrix
+    # is not necessarily perfectly symmetric but only up to
+    # round-off errors. We wrap it here to avoid issues with the
+    # factorization.
+    # TODO: This can be made more efficient, maybe by implementing a
+    #       new interface in SummationByPartsOperators.jl to allow
+    #       evaluating the product of a derivative operator with the
+    #       corresponding mass matrix directly.
+    @.. D2.b = h^3 / 3
+    copyto!(A, D2)
+    A = Diagonal(M_h) - mass_matrix(D1) * A
+    A[begin, :] .= 0
+    A[:, begin] .= 0
+    A[begin, begin] = 1
+    A[end, :] .= 0
+    A[:, end] .= 0
+    A[end, end] = 1
+    return Symmetric(A)
+end
+
+function rhs!(dq, q, t, mesh,
+              equations::SerreGreenNaghdiEquations1D,
```

Comment:
> For a variable bathymetry, we will need to discretize more terms, and I would like to check first by hand how they fit together for energy conservation. Thus, I have restricted this PR to the flat case and would like to implement variable bathymetry in another PR (also to keep the number of changes per PR small enough for review).

Sounds good to me. In this case, would it make sense to also restrict the `rhs!` to only flat bottom?
```suggestion
              equations::SerreGreenNaghdiEquations1D{BathymetryFlat},
```